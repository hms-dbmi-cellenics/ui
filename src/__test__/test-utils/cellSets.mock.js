const mockCellSets = {
  initialLoadPending: false,
  error: false,
  loading: false,
  accessible: true,
  hierarchy: [
    {
      key: 'louvain',
      children: [{ key: 'cluster-a' }, { key: 'cluster-b' }, { key: 'cluster-c' }],
      cellIds: new Set(),
    },
    {
      key: 'sample',
      children: [{ key: 'sample-1' }, { key: 'sample-2' }],
      cellIds: new Set(),
    },
    {
      key: 'scratchpad',
      children: [{ key: 'scratchpad-a' }],
      cellIds: new Set(),
    },
  ],
  properties: {
    'cluster-a': {
      name: 'cluster a',
      key: 'cluster-a',
      cellIds: new Set([0, 1, 6]),
      color: '#01FFFF',
    },
    'cluster-b': {
      name: 'cluster b',
      key: 'cluster-b',
      cellIds: new Set([2, 3]),
      color: '#23FFFF',
    },
    'cluster-c': {
      name: 'cluster c',
      key: 'cluster-c',
      cellIds: new Set([4, 5]),
      color: '#45FFFF',
    },
    'sample-1': {
      name: 'Sample 1',
      key: 'sample-1',
      cellIds: new Set([0, 1, 2]),
      color: '#012FFF',
    },
    'sample-2': {
      name: 'Sample 2',
      key: 'sample-2',
      cellIds: new Set([3, 4, 5]),
      color: '#345FFF',
    },
    'scratchpad-a': {
      cellIds: new Set(['5']),
      key: 'scratchpad-a',
      name: 'New Cluster',
      color: '#5FFFFF',
    },
    louvain: {
      cellIds: new Set(),
      name: 'Louvain clusters',
      key: 'louvain',
      type: 'cellSets',
      rootNode: true,
    },
    sample: {
      cellIds: new Set([6, 7, 8]),
      name: 'Samples',
      key: 'sample',
      type: 'metadataCategorical',
      rootNode: true,
    },
    scratchpad: {
      cellIds: new Set(),
      name: 'Scratchpad',
      key: 'scratchpad',
      type: 'cellSets',
      rootNode: true,
    },
  },
};

export {
  // eslint-disable-next-line import/prefer-default-export
  mockCellSets,
};
