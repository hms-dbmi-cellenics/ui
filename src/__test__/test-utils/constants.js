export default {
  EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperiment',
  CLONED_EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperclone',
  PROJECT_WITH_SAMPLES_UUID: '380c53b4-mock-test-8091-4c82336d6d49',
  PROJECT_WITHOUT_SAMPLE_UUID: '4761594b-mock-test-ba89-c2c812d39bb5',
  EXPERIMENT_NAME: 'Test Experiment',
  PROJECT_ID: 'test24b6-8600-test-mock-63822d2fmock',
  SAMPLE_ID: 'test9188-d682-test-mock-cb6d644cmock',
  API_ENDPOINT: /^http?:\/\/localhost:3000.*$/,
  S3_ENDPOINT: 'http://mock.s3.amazonaws.com',
  // pragma: allowlist secret
  JWT_TOKEN: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6Ikpfake.eyJzdWIiOiIxMjM0NTY3ODkwIiwifakeZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTfakeM5MDIyfQ.fakexwRfakeKKF2QT4fwpMeJf36POk6yJV_adQsfake',
  MOCK_DATETIME: '0000-00-00T00:00:00.000Z',
  ETAG: '380c53b4-mock-etag-8091-4c82336d6d49',
};
