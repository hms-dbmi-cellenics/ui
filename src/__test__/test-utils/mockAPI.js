import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

import fake from '__test__/test-utils/constants';

import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => (
  Promise.resolve({
    status: code,
    body: JSON.stringify(body),
  })
);

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerDataResult = (data) => Promise.resolve(_.cloneDeep(data));

const fetchWorkMock = (
  mockedResults,
) => ((experimentId, body) => {
  if (typeof mockedResults[body.name] === 'function') {
    return mockedResults[body.name]();
  }

  return workerDataResult(mockedResults[experimentId]);
});

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments$: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples$`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples$': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
  'experiments/clone$': () => promiseResponse(
    JSON.stringify(fake.CLONED_EXPERIMENT_ID),
  ),
  [`/v2/workRequest/${experimentId}`]: () => statusResponse(200, 'OK'),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(Object.keys(apiMapping), (matcherStr) => new RegExp(matcherStr).test(path));

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
  workerDataResult,
  fetchWorkMock,
};
