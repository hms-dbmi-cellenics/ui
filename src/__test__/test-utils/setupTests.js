import '@testing-library/jest-dom/extend-expect';
import preloadAll from 'jest-next-dynamic';

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

// This is needed, because the unit tests don't register the Vitessce imports
jest.mock('components/data-exploration/DynamicVitessceWrappers', () => ({
  Scatterplot: () => 'Mocked Scatterplot',
  Heatmap: () => 'Mocked Heatmap',
  Spatial: () => 'Mocked Spatial',
}));

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});
