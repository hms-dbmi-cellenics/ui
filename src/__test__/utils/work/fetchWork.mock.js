const mockGenesListData = ['A', 'B', 'C', 'D'];

// const mockGeneExpressionData = {
//   A: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//   },
//   B: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.702857143,
//       stdev: 2.551115536,
//       expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.702857143,
//       stdev: 2.551115536,
//       expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
//     },
//   },
//   C: {
//     rawExpression: {
//       min: 0,
//       max: 3.4,
//       mean: 1.68,
//       stdev: 2.141525936,
//       expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 3.4,
//       mean: 1.68,
//       stdev: 2.141525936,
//       expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
//     },
//   },
//   D: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//   },
//   E: { hello: 'world' },
// };

// const mockCacheGet = jest.fn((x) => {
//   const mockCacheContents = {
//     fd3161a878f67ebf54018720cffd6a66: 'A', // pragma: allowlist secret
//     '10250a11679234110a1c260d6fd81d3c': 'B', // pragma: allowlist secret
//     f5c957411a28de68f35e1f5c8a29da7e: 'C', // pragma: allowlist secret
//     '33faa711a94a2028b5bae1778126aec0': 'E', // pragma: allowlist secret
//   };

//   if (x in mockCacheContents) {
//     return mockGeneExpressionData[mockCacheContents[x]];
//   }

//   return null;
// });

const mockCacheGet = jest.fn();
const mockCacheSet = jest.fn();
const mockCacheRemove = jest.fn();

const mockFetchWorkRequest = jest.fn(() => true);

const mockSeekFromS3 = jest.fn();

const mockCacheModule = {
  get: jest.fn((x) => mockCacheGet(x)),
  set: jest.fn((key, val) => mockCacheSet(key, val)),
  _remove: jest.fn((key) => mockCacheRemove(key)),
};

const mockFetchWorkModule = {
  __esModule: true,
  fetchWork: mockFetchWorkRequest,
};

const mockQcPipelineStartDate = '2021-01-01T01:01:01.000Z';

export {
  mockGenesListData, mockCacheGet, mockCacheSet,
  mockFetchWorkRequest, mockSeekFromS3,
  mockQcPipelineStartDate, mockCacheModule, mockFetchWorkModule,
};
