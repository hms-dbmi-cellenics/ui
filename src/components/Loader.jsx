import React from 'react';
import PropTypes from 'prop-types';
import { ClipLoader, BounceLoader } from 'react-spinners';
import { Typography } from 'antd';
import useSWR from 'swr';

import fetchAPI from 'utils/http/fetchAPI';

const { Text } = Typography;

const slowLoad = () => (
  <>
    <div style={{ padding: 25 }}>
      <center>
        <BounceLoader
          size={50}
          color='#8f0b10'
          css={{ display: 'block' }}
        />
      </center>
    </div>
    <p>
      <Text>
        This will take a few minutes...
      </Text>
    </p>
    <p>
      <Text type='secondary'>
        We&apos;re setting up your analysis after a period of inactivity. Please wait.
      </Text>
    </p>
  </>
);

const fastLoad = (message) => (
  <>
    <div style={{ padding: 25 }}>
      <ClipLoader
        size={50}
        color='#8f0b10'
      />
    </div>
    <p>
      <Text>
        {message || "We're getting your data ..."}
      </Text>
    </p>
  </>
);

const Loader = ({ experimentId }) => {
  const { data: workerStatus } = useSWR(
    () => (experimentId ? `/v1/experiments/${experimentId}/backendStatus` : null),
    fetchAPI,
  );

  if (!workerStatus) {
    return (
      <div>
        {fastLoad()}
      </div>
    );
  }

  const { worker: { started, ready } } = workerStatus;
  if (started && ready) {
    return (
      <div>
        {fastLoad()}
      </div>
    );
  }

  return (
    <div>
      {slowLoad()}
    </div>
  );
};

Loader.propTypes = {
  experimentId: PropTypes.string.isRequired,
};

export default Loader;
export { fastLoad, slowLoad };
