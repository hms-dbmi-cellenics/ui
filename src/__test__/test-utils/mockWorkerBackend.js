import * as socketConnectionMocks from 'utils/socketConnection';
import * as fetchWorkResponseMocks from 'utils/work/fetchWork';

import SocketMock from 'socket.io-mock';

const socketMock = new SocketMock();

jest.mock('utils/work/fetchWork', () => {
  const mockSeekFromS3 = jest.fn();
  const originalModule = jest.requireActual('utils/work/fetchWork');

  return {
    __esModule: true, // Use it when dealing with esModules
    ...originalModule,
    fetchWork: mockSeekFromS3,
  };
});

jest.mock('aws-amplify', () => ({
  configure: () => ({
    Storage: {
      AWSS3: {
        bucket: 'biomage-originals-test',
      },
    },
  }),
  Storage: {
    get: async (ETag) => `http://mock.s3.amazonaws.com/${ETag}`,
  },
  Auth: {
    federatedSignIn: jest.fn(),
    signOut: jest.fn(),
    currentAuthenticatedUser: jest.fn().mockImplementation(async () => ({
      username: 'mockuser',
      attributes: {
        email: 'mock@user.name',
        name: 'Mocked User',
      },
    })),
    currentSession: () => ({
      getIdToken: (() => ({
        getJwtToken: () => 'fakeJwtToken',
      })),
    }),
  },
  Hub: {
    listen: jest.fn(),
  },
}));

jest.mock('utils/socketConnection', () => {
  const mockEmit = jest.fn();
  const mockOn = jest.fn();
  return {
    __esModule: true,
    default: new Promise((resolve) => {
      resolve({ emit: mockEmit, on: mockOn, id: '1234' });
    }),
    mockEmit,
    mockOn,
  };
});

// To mock worker response, modify the response returned from the API
// Set to null to force fetch from API
fetchWorkResponseMocks.fetchWork.mockImplementation(() => null);

// Set up socket emitter mock
socketConnectionMocks.mockEmit.mockImplementation((workRequestType, requestBody) => {
  const responseBody = { response: { error: false } };

  // After emitting, send reply to listener
  socketMock.socketClient.emit(`WorkResponse-${requestBody.ETag}`, responseBody);
});

// Set up socket listener mock
socketConnectionMocks.mockOn.mockImplementation((channel, f) => {
  socketMock.on(channel, (responseBody) => f(responseBody));
});
