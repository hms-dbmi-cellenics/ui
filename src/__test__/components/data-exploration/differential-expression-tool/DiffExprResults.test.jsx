import React from 'react';
import { mount, shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Empty } from 'antd';
import _ from 'lodash';
import waitForActions from 'redux-mock-store-await-actions';
import AdvancedFilteringModal from 'components/data-exploration/differential-expression-tool/AdvancedFilteringModal';

import DiffExprResults from 'components/data-exploration/differential-expression-tool/DiffExprResults';
import { DIFF_EXPR_LOADING, DIFF_EXPR_LOADED, DIFF_EXPR_ORDERING_SET } from 'redux/actionTypes/differentialExpression';

import { mockCellSets } from '__test__/test-utils/cellSets.mock';

import Loader from 'components/Loader';
import { fetchWork } from 'utils/work/fetchWork';

jest.mock('utils/getTimeoutForWorkerTask', () => ({
  __esModule: true, // this property makes it work
  default: () => 60,
}));

jest.mock('utils/work/fetchWork', () => ({
  __esModule: true, // this property makes it work
  fetchWork: jest.fn(() => new Promise((resolve) => resolve({
    rows: [
      {
        p_val: 1.496, p_val_adj: 1.647, logFC: -1.427, gene_names: 'A', auc: '0.1', pct_1: '100', pct_2: '100',
      },
      {
        p_val: 2.496, p_val_adj: 2.647, logFC: -2.427, gene_names: 'B', auc: '0.2', pct_1: '90', pct_2: '90',
      },
      {
        p_val: 3.496, p_val_adj: 3.647, logFC: -3.427, gene_names: 'C', auc: '0.3', pct_1: '80', pct_2: '80',
      },
      {
        p_val: 4.496, p_val_adj: 4.647, logFC: -4.427, gene_names: 'D', auc: '0.4', pct_1: '70', pct_2: '70',
      },
      {
        p_val: 5.496, p_val_adj: 5.647, logFC: -5.427, gene_names: 'E', auc: '0.5', pct_1: '60', pct_2: '60',
      },
    ],
    total: 500,
  }))),
}));

const mockStore = configureMockStore([thunk]);

const experimentId = '1234';

const backendStatus = {
  [experimentId]: {
    status: {
      pipeline: {
        startDate: '2021-01-01T00:00:00',
      },
    },
  },
};

const mockGeneExpressionData = [
  {
    p_val: 1.496, p_val_adj: 1.647, logFC: -1.427, gene_names: 'A', auc: '0.1', pct_1: '100', pct_2: '100', Gene: 'EN123123123123123',
  },
  {
    p_val: 2.496, p_val_adj: 2.647, logFC: -2.427, gene_names: 'B', auc: '0.2', pct_1: '90', pct_2: '90', Gene: 'EN123123123123123',
  },
  {
    p_val: 3.496, p_val_adj: 3.647, logFC: -3.427, gene_names: 'C', auc: '0.3', pct_1: '80', pct_2: '80', Gene: 'EN123123123123123',
  },
  {
    p_val: 4.496, p_val_adj: 4.647, logFC: -4.427, gene_names: 'D', auc: '0.4', pct_1: '70', pct_2: '70', Gene: 'EN123123123123123',
  },
  {
    p_val: 5.496, p_val_adj: 5.647, logFC: -5.427, gene_names: 'E', auc: '0.5', pct_1: '60', pct_2: '60', Gene: 'EN123123123123123',
  },
];

const resultState = {
  genes: {
    selected: [],
  },
  cellInfo: {
    focus: {
      store: 'genes',
      key: 'C',
    },
  },
  cellSets: mockCellSets,
  differentialExpression: {
    properties: {
      data: mockGeneExpressionData,
      loading: false,
      error: false,
      total: 5,
    },
    comparison: {
      type: 'between',
      group: {
        between: {
          cellSet: 'louvain/cluster-a',
          compareWith: 'louvain/cluster-b',
          basis: 'scratchpad/scratchpad-a',
        },
      },
      advancedFilters: [],
    },
  },
  backendStatus,
};

// State with less gene expression fields
const partialState = _.cloneDeep(resultState);
const partialGeneExpData = mockGeneExpressionData.map((data) => {
  const { pct_1, pct_2, ...remaining } = data;
  return remaining;
});
partialState.differentialExpression.properties.data = partialGeneExpData;

// State with more gene expression fields
const extraState = _.cloneDeep(resultState);
const extraGeneExpData = mockGeneExpressionData.map((data, idx) => {
  const newData = {
    ...data,
    extra_field: idx,
    more_extra_field: idx + 1,
  };

  return newData;
});
extraState.differentialExpression.properties.data = extraGeneExpData;

// State with no results
const noResultState = {
  ...resultState,
  differentialExpression: {
    ...resultState.differentialExpression,
    properties: {
      ...resultState.differentialExpression.properties,
      data: [],
      loading: false,
      error: false,
      total: 0,
    },
  },
};

const withResultStore = mockStore(resultState);
const partialResultStore = mockStore(partialState);
const extraResultStore = mockStore(extraState);
const noResultStore = mockStore(noResultState);

describe('DiffExprResults', () => {
  it('renders correctly', () => {
    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults
          experimentId={experimentId}
          onGoBack={jest.fn()}
          width={100}
          height={200}
        />
      </Provider>,
    );

    const table = component.find('Table');
    const spin = component.find(Loader);
    expect(spin.length).toEqual(0);
    expect(table.length).toEqual(1);
    expect(table.getElement().props.columns.length).toEqual(7);
    expect(table.getElement().props.columns[0].key).toEqual('lookup');
    expect(table.getElement().props.columns[1].key).toEqual('gene_names');
    expect(table.getElement().props.columns[2].key).toEqual('logFC');
    expect(table.getElement().props.columns[3].key).toEqual('p_val_adj');
    expect(table.getElement().props.columns[4].key).toEqual('pct_1');
    expect(table.getElement().props.columns[5].key).toEqual('pct_2');
    expect(table.getElement().props.columns[6].key).toEqual('auc');

    expect(table.getElement().props.dataSource.length).toEqual(5);
    expect(table.getElement().props.data.length).toEqual(5);
  });

  it('is sorted by descending logFC by default', () => {
    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults
          experimentId={experimentId}
          onGoBack={jest.fn()}
          width={100}
          height={200}
        />
      </Provider>,
    );

    const table = component.find('Table');
    expect(table.getElement().props.columns[1].sortOrder).toEqual(null);
    expect(table.getElement().props.columns[2].sortOrder).toEqual('descend');
    expect(table.getElement().props.columns[3].sortOrder).toEqual(null);
    expect(table.getElement().props.columns[4].sortOrder).toEqual(null);
    expect(table.getElement().props.columns[5].sortOrder).toEqual(null);
    expect(table.getElement().props.columns[6].sortOrder).toEqual(null);
  });

  it('fetches all the genes', async () => {
    const newPagination = {
      current: 1,
      pageSize: 4,
      showSizeChanger: true,
      total: mockGeneExpressionData.length,
    };

    const newSorter = {
      column: {
        dataIndex: 'gene_names',
        key: 'gene_names',
      },
      render: jest.fn(),
      columnKey: 'gene_names',
      field: 'gene_names',
      order: 'ascend',
    };

    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults
          experimentId={experimentId}
          onGoBack={jest.fn()}
          width={100}
          height={200}
        />
      </Provider>,
    );

    const table = component.find('Table');

    act(() => {
      table.getElement().props.onChange(newPagination, {}, newSorter);
    });

    // // Wait for side-effect to propagate (properties loading and loaded).
    await waitForActions(withResultStore, [DIFF_EXPR_ORDERING_SET, DIFF_EXPR_LOADING, DIFF_EXPR_LOADED]);
    const entries = component.find('.ant-table-tbody').children();

    expect(fetchWork).toHaveBeenCalledWith(
      '1234',
      {
        cellSet: 'cluster-a',
        compareWith: 'cluster-b',
        basis: 'scratchpad-a',
        comparisonType: 'between',
        experimentId: '1234',
        name: 'DifferentialExpression',
      },
      withResultStore.getState,
      {
        extras: {
          pagination: {
            limit: 1000000, offset: 0, orderBy: 'logFC', orderDirection: 'DESC', responseKey: 0,
          },
        },
        timeout: 60,
      },
    );
    expect(entries).toHaveLength(6);
    expect(withResultStore.getActions()[1]).toMatchSnapshot();
    expect(withResultStore.getActions()[2]).toMatchSnapshot();
  });

  it('displays the correct genes when searching', () => {
    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults
          experimentId={experimentId}
          onGoBack={jest.fn()}
          width={100}
          height={200}
        />
      </Provider>,
    );

    const searchBox = component.find('.ant-input');

    searchBox.simulate('change', { target: { value: 'A' } });

    const entries = component.find('.ant-table-tbody').children();

    expect(entries.at(1).text()).toEqual('A-1.4271.6471001000.1');
    expect(entries).toHaveLength(2);
  });

  it('Having a focused gene triggers focused view for `eye` button.', () => {
    // Redefine store from `beforeEach`.
    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults
          experimentId={experimentId}
          onGoBack={jest.fn()}
          width={100}
          height={200}
        />
      </Provider>,
    );

    const table = component.find('Table');

    table.getElement().props.data.forEach((row) => {
      const lookupComponent = mount(
        <Provider store={withResultStore}>
          {row.lookup}
        </Provider>,
      );

      const focusButtonTooltip = lookupComponent.find('FocusButton Tooltip');

      if (row.gene_names === 'C') {
        expect(focusButtonTooltip.props().title).toContain('Hide');
      } else {
        expect(focusButtonTooltip.props().title).toContain('Show');
      }

      lookupComponent.unmount();
    });
  });

  it('Show comparison settings button works.', () => {
    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults
          experimentId={experimentId}
          onGoBack={jest.fn()}
          width={100}
          height={200}
        />
      </Provider>,
    );
    const button = component.find('#settingsButton').first();
    expect(button.text()).toContain('Show');
    button.simulate('click');
    expect(button.text()).toContain('Hide');

    const div = component.find('#settingsText');
    // Should display name of cluster instead of ID
    expect(div.text()).toEqual('cluster a vs. cluster b in New Cluster');
    button.simulate('click');
    expect(button.childAt(0).text()).toEqual('Show settings');
    expect(!div);
  });

  it('Does not show loading indicator if there is no data returned', () => {
    const component = mount(
      <Provider store={noResultStore}>
        <DiffExprResults
          experimentId={experimentId}
        />
      </Provider>,
    );

    const spin = component.find(Loader);
    const empty = component.find(Empty);

    // There should be no loader
    expect(spin.length).toEqual(0);

    // Expect table to contain Empty component
    expect(empty.length).toEqual(1);
  });

  it('Advanced filter button opens and closes the modal', async () => {
    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults
          experimentId={experimentId}
          onGoBack={jest.fn()}
          width={100}
          height={200}
        />
      </Provider>,
    );
    const buttons = component.find('Button');
    expect(buttons.at(2).text()).toEqual('Advanced filtering');

    // opening the modal
    buttons.at(2).simulate('click');
    expect(component.find(AdvancedFilteringModal).length).toEqual(1);

    // Adding a filter and applying it
    const dropdown = component.find('Dropdown');
    const menuInstance = shallow(dropdown.props().overlay);
    menuInstance.at(0).simulate('click');
    await waitForActions(withResultStore, [DIFF_EXPR_ORDERING_SET, DIFF_EXPR_LOADING, DIFF_EXPR_LOADED]);

    // closing the modal
    const closeButton = component.find('.ant-modal-close');
    closeButton.simulate('click');
    expect(component.find(AdvancedFilteringModal).length).toEqual(0);
  });

  it('Pathway analysis button opens and closes the modal and dispatches loadDifferentialExpression', async () => {
    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults
          experimentId={experimentId}
          onGoBack={jest.fn()}
          width={100}
          height={200}
        />
      </Provider>,
    );

    // On clicking LaunchPathwayAnalysisModal button
    const buttons = component.find('span[children="Pathway analysis"]');
    expect(buttons.at(0).text()).toEqual('Pathway analysis');
    buttons.at(0).simulate('click');

    // Shows the modal
    expect(component.find('LaunchPathwayAnalysisModal').length).toEqual(1);

    // closing the modal
    const closeButton = component.find('.ant-modal-close');
    closeButton.simulate('click');
    expect(component.find('LaunchPathwayAnalysisModal').length).toEqual(0);
  });

  it('Columns without corresponding data are not shown', async () => {
    const component = mount(
      <Provider store={partialResultStore}>
        <DiffExprResults
          experimentId={experimentId}
          onGoBack={jest.fn()}
          width={100}
          height={200}
        />
      </Provider>,
    );

    const table = component.find('Table');
    expect(table.getElement().props.columns.length).toEqual(5);
    expect(table.getElement().props.columns[0].key).toEqual('lookup');
    expect(table.getElement().props.columns[1].key).toEqual('gene_names');
    expect(table.getElement().props.columns[2].key).toEqual('logFC');
    expect(table.getElement().props.columns[3].key).toEqual('p_val_adj');
    expect(table.getElement().props.columns[4].key).toEqual('auc');
  });

  it('Data without corresponding columns are not shown', async () => {
    const component = mount(
      <Provider store={extraResultStore}>
        <DiffExprResults
          experimentId={experimentId}
          onGoBack={jest.fn()}
          width={100}
          height={200}
        />
      </Provider>,
    );

    const table = component.find('Table');

    expect(table.getElement().props.columns.length).toEqual(7);
    expect(table.getElement().props.columns[0].key).toEqual('lookup');
    expect(table.getElement().props.columns[1].key).toEqual('gene_names');
    expect(table.getElement().props.columns[2].key).toEqual('logFC');
    expect(table.getElement().props.columns[3].key).toEqual('p_val_adj');
    expect(table.getElement().props.columns[4].key).toEqual('pct_1');
    expect(table.getElement().props.columns[5].key).toEqual('pct_2');
    expect(table.getElement().props.columns[6].key).toEqual('auc');
  });

  it('The export as CSV alert opens and closes properly', async () => {
    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults
          experimentId={experimentId}
          onGoBack={jest.fn()}
          width={100}
          height={200}
        />
      </Provider>,
    );

    // Clicking the CSV button opens the modal
    const csvButton = component.find('span[children="Export as CSV"]');
    expect(csvButton.length).toEqual(1);

    act(() => {
      csvButton.simulate('click');
    });
    component.update();

    const csvModal = component.find('Alert');
    expect(csvModal.length).toEqual(1);

    // Clicking the close button closes the CSV modal
    const closeCsvModalButton = csvModal.find('button');
    expect(closeCsvModalButton.length).toEqual(1);

    act(() => {
      closeCsvModalButton.simulate('click');
    });
    component.update();

    // Expect CSV modal to not be shown anymore
    expect(component.find('Alert').length).toEqual(0);
  });
});
