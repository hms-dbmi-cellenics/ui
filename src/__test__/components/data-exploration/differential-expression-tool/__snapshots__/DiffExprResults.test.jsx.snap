// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffExprResults fetches all the genes 1`] = `
Object {
  "payload": Object {
    "advancedFilters": Array [],
    "experimentId": "1234",
  },
  "type": "diffExpr/loading",
}
`;

exports[`DiffExprResults fetches all the genes 2`] = `
Object {
  "payload": Object {
    "comparisonGroup": Object {
      "basis": "scratchpad/scratchpad-a",
      "cellSet": "louvain/cluster-a",
      "compareWith": "louvain/cluster-b",
    },
    "comparisonType": "between",
    "data": Array [
      Object {
        "auc": "0.1",
        "gene_names": "A",
        "logFC": -1.427,
        "p_val": 1.496,
        "p_val_adj": 1.647,
        "pct_1": "100",
        "pct_2": "100",
      },
      Object {
        "auc": "0.2",
        "gene_names": "B",
        "logFC": -2.427,
        "p_val": 2.496,
        "p_val_adj": 2.647,
        "pct_1": "90",
        "pct_2": "90",
      },
      Object {
        "auc": "0.3",
        "gene_names": "C",
        "logFC": -3.427,
        "p_val": 3.496,
        "p_val_adj": 3.647,
        "pct_1": "80",
        "pct_2": "80",
      },
      Object {
        "auc": "0.4",
        "gene_names": "D",
        "logFC": -4.427,
        "p_val": 4.496,
        "p_val_adj": 4.647,
        "pct_1": "70",
        "pct_2": "70",
      },
      Object {
        "auc": "0.5",
        "gene_names": "E",
        "logFC": -5.427,
        "p_val": 5.496,
        "p_val_adj": 5.647,
        "pct_1": "60",
        "pct_2": "60",
      },
    ],
    "experimentId": "1234",
    "total": 500,
  },
  "type": "diffExpr/loaded",
}
`;
